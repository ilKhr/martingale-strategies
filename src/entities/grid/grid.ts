import {
  longCalculation,
  LongCalculationPayload,
  shortCalculation,
  ShortCalculationPayload,
} from "../../utils/formulas";
import {
  BaseOrderActionsCallbacks,
  createAndAddOrder,
  LimitOrder,
  Order,
} from "../order/order";

import { Percent, TradingAlgorithm } from "../../types";

export type Grid = {
  orders: Order[];
  configuration: {
    tradingAlgorithm: TradingAlgorithm;
    countOrders: number;
    profit: number;
    overlap: number;
    startPrice: string;
    stopLoss?: number;
  };
};

export type CreateGridParams = {
  countOrders: number;
  deposit: string;
  martingale: Percent;
  overlap: Percent;
  currencyPrice: string;
  profit: Percent;
  tradingAlgorithm: TradingAlgorithm;
  startPrice: string;
  stopLoss?: number;
};

export type CalculateOrderGridType = (
  params: {
    countOrders: number;
    deposit: string;
    martingale: Percent;
    overlap: Percent;
    currencyPrice: string;
    profit: Percent;
    tradingAlgorithm: TradingAlgorithm;
  },
  createOrderCallback?: BaseOrderActionsCallbacks["createOrder"]
) => LimitOrder[];

export const calculateOrderGrid: CalculateOrderGridType = (
  params,
  createOrderCallback
) => {
  const orderGrid: LimitOrder[] = [];

  for (
    let sequenceIndexInSide = 0;
    sequenceIndexInSide < params.countOrders;
    sequenceIndexInSide++
  ) {
    const isLongOrder = params.tradingAlgorithm === "LONG";

    const calculationArgument:
      | LongCalculationPayload
      | ShortCalculationPayload = Object.assign({}, params, {
      sequenceIndexInSide: sequenceIndexInSide,
    });

    const { priceWithIntent, quantity } = isLongOrder
      ? longCalculation(calculationArgument)
      : shortCalculation(calculationArgument);

    // TODO: calculate price with exchange filter
    // TODO: calculate quantity with exchange filter

    createAndAddOrder(
      {
        price: priceWithIntent.toString(),
        quantity: quantity.toString(),
        side: isLongOrder ? "BUY" : "SELL",
      },
      { orders: orderGrid },
      createOrderCallback
    );
  }

  return orderGrid;
};
