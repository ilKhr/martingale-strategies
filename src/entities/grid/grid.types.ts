import { ID, Percent, Side, TradingAlgorithm } from "../../types";
import { Order } from "../order/order.types";

export type Grid = {
  orders: Order[];
  configuration: {
    tradingAlgorithm: TradingAlgorithm;
    countOrders: number;
    profit: number;
    overlap: number;
    startPrice: string;
    stopLoss?: number;
  };
  status: "active" | "done" | "canceled" | "error";
};

export type CreateGridParams = {
  countOrders: number;
  deposit: string;
  martingale: Percent;
  overlap: Percent;
  currencyPrice: string;
  profit: Percent;
  tradingAlgorithm: TradingAlgorithm;
  startPrice: string;
  stopLoss?: number;
};

export type CalculateOrderGridType = (params: {
  countOrders: number;
  deposit: string;
  martingale: Percent;
  overlap: Percent;
  currencyPrice: string;
  profit: Percent;
  tradingAlgorithm: TradingAlgorithm;
}) => {
  id: ID;
  price: string;
  quantity: string;
  side: Side;
  status: "active";
  sequenceIndexInSide: number;
}[];
