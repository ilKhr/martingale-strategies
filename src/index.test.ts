import { describe, vi, test, expect, afterEach } from "vitest";
import { MartingaleStrategies } from "./";
import { Grid } from "./entities/grid/grid";

const createOrder = vi.fn();
const cancelOrder = vi.fn();
const createOcoOrder = vi.fn();
const markOrderAsDone = vi.fn();

afterEach(() => {
  vi.clearAllMocks();
});

describe("MartingaleStrategies", () => {
  test("Work correctly with onOrderDone", () => {
    const configuration = {
      countOrders: 10,
      currencyPrice: "100",
      deposit: "1000",
      martingale: 5,
      overlap: 50,
      profit: 1,
      currencyPriceInStart: "100",
      tradingAlgorithm: "LONG",
      stopLoss: 5,
    } as const;

    const instance = MartingaleStrategies({
      createOrder,
      cancelOrder,
      createOcoOrder,
      markOrderAsDone,
    });

    const grid = instance.createGrid(configuration);

    const first = instance.onOrderDone({
      grid: {
        configuration: grid.configuration,
        orders: grid.orders,
      },
      triggerOrderId: 1,
    });

    expect(first).toEqual({
      grid: {
        configuration: {
          countOrders: 10,
          profit: 1,
          tradingAlgorithm: "LONG",
          overlap: 50,
          currencyPriceInStart: "100",
          stopLoss: 5,
        },
        orders: [
          {
            id: 1,
            status: "done",
            sequenceIndexInSide: 0,
            price: "95",
            quantity: "0.83689026279428100526",
            side: "BUY",
            type: "LIMIT",
          },
          {
            id: 2,
            status: "active",
            sequenceIndexInSide: 1,
            price: "90",
            quantity: "0.92755337459699478083",
            side: "BUY",
            type: "LIMIT",
          },
          {
            id: 3,
            status: "active",
            sequenceIndexInSide: 2,
            price: "85",
            quantity: "1.03122110469901184457",
            side: "BUY",
            type: "LIMIT",
          },
          {
            id: 4,
            status: "active",
            sequenceIndexInSide: 3,
            price: "80",
            quantity: "1.1504560449298350891",
            side: "BUY",
            type: "LIMIT",
          },
          {
            id: 5,
            status: "active",
            sequenceIndexInSide: 4,
            price: "75",
            quantity: "1.28851077032141529979",
            side: "BUY",
            type: "LIMIT",
          },
          {
            id: 6,
            status: "active",
            sequenceIndexInSide: 5,
            price: "70",
            quantity: "1.44957461661159221227",
            side: "BUY",
            type: "LIMIT",
          },
          {
            id: 7,
            status: "active",
            sequenceIndexInSide: 6,
            price: "65",
            quantity: "1.63913437416849273233",
            side: "BUY",
            type: "LIMIT",
          },
          {
            id: 8,
            status: "active",
            sequenceIndexInSide: 7,
            price: "60",
            quantity: "1.86451535061666048303",
            side: "BUY",
            type: "LIMIT",
          },
          {
            id: 9,
            status: "active",
            sequenceIndexInSide: 8,
            price: "55",
            quantity: "2.13571758343362928056",
            side: "BUY",
            type: "LIMIT",
          },
          {
            id: 10,
            status: "active",
            sequenceIndexInSide: 9,
            price: "50",
            quantity: "2.46675380886584181904",
            side: "BUY",
            type: "LIMIT",
          },
          {
            id: 11,
            status: "active",
            sequenceIndexInSide: 0,
            price: "95.95",
            quantity: "0.83689026279428100526",
            side: "SELL",
            type: "LIMIT",
          },
        ],
      },
      isCycleOver: false,
    });

    expect(markOrderAsDone).toBeCalledTimes(1);
    expect(createOrder).toBeCalledTimes(11);
    expect(cancelOrder).toBeCalledTimes(0);
    expect(createOcoOrder).toBeCalledTimes(0);
    expect(markOrderAsDone).toBeCalledWith({
      id: 1,
      status: "done",
      sequenceIndexInSide: 0,
      price: "95",
      quantity: "0.83689026279428100526",
      side: "BUY",
      type: "LIMIT",
    });
  });

  test("Work correctly with x10 onOrderDone", () => {
    const configuration = {
      countOrders: 10,
      currencyPrice: "100",
      deposit: "1000",
      martingale: 5,
      overlap: 50,
      profit: 1,
      currencyPriceInStart: "100",
      tradingAlgorithm: "LONG",
      stopLoss: 5,
    } as const;

    const instance = MartingaleStrategies({
      createOrder,
      cancelOrder,
      createOcoOrder,
      markOrderAsDone,
    });

    const grid = instance.createGrid(configuration);

    let result:
      | {
          isCycleOver: boolean;
          grid: Grid;
        }
      | undefined = undefined;

    for (let i = 1; i < 11; i++) {
      result = instance.onOrderDone({
        grid: result?.grid || grid,
        triggerOrderId: i,
      });
    }

    expect(result).toMatchSnapshot();

    expect(markOrderAsDone).toBeCalledTimes(10);
    expect(createOrder).toBeCalledTimes(19);
    expect(cancelOrder).toBeCalledTimes(9);
    expect(createOcoOrder).toBeCalledTimes(1);
    expect(markOrderAsDone).toBeCalledWith({
      id: 1,
      status: "done",
      sequenceIndexInSide: 0,
      price: "95",
      quantity: "0.83689026279428100526",
      side: "BUY",
      type: "LIMIT",
    });
  });
});
