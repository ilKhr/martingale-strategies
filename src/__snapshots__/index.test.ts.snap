// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MartingaleStrategies > Work correctly with x10 onOrderDone 1`] = `
{
  "grid": {
    "configuration": {
      "countOrders": 10,
      "currencyPriceInStart": "100",
      "overlap": 50,
      "profit": 1,
      "stopLoss": 5,
      "tradingAlgorithm": "LONG",
    },
    "orders": [
      {
        "id": 1,
        "oco": {
          "isOco": false,
        },
        "price": "95",
        "quantity": "0.83689026279428100526",
        "sequenceIndexInSide": 0,
        "side": "BUY",
        "status": "done",
        "type": "LIMIT",
      },
      {
        "id": 2,
        "oco": {
          "isOco": false,
        },
        "price": "90",
        "quantity": "0.92755337459699478083",
        "sequenceIndexInSide": 1,
        "side": "BUY",
        "status": "done",
        "type": "LIMIT",
      },
      {
        "id": 3,
        "oco": {
          "isOco": false,
        },
        "price": "85",
        "quantity": "1.03122110469901184457",
        "sequenceIndexInSide": 2,
        "side": "BUY",
        "status": "done",
        "type": "LIMIT",
      },
      {
        "id": 4,
        "oco": {
          "isOco": false,
        },
        "price": "80",
        "quantity": "1.1504560449298350891",
        "sequenceIndexInSide": 3,
        "side": "BUY",
        "status": "done",
        "type": "LIMIT",
      },
      {
        "id": 5,
        "oco": {
          "isOco": false,
        },
        "price": "75",
        "quantity": "1.28851077032141529979",
        "sequenceIndexInSide": 4,
        "side": "BUY",
        "status": "done",
        "type": "LIMIT",
      },
      {
        "id": 6,
        "oco": {
          "isOco": false,
        },
        "price": "70",
        "quantity": "1.44957461661159221227",
        "sequenceIndexInSide": 5,
        "side": "BUY",
        "status": "done",
        "type": "LIMIT",
      },
      {
        "id": 7,
        "oco": {
          "isOco": false,
        },
        "price": "65",
        "quantity": "1.63913437416849273233",
        "sequenceIndexInSide": 6,
        "side": "BUY",
        "status": "done",
        "type": "LIMIT",
      },
      {
        "id": 8,
        "oco": {
          "isOco": false,
        },
        "price": "60",
        "quantity": "1.86451535061666048303",
        "sequenceIndexInSide": 7,
        "side": "BUY",
        "status": "done",
        "type": "LIMIT",
      },
      {
        "id": 9,
        "oco": {
          "isOco": false,
        },
        "price": "55",
        "quantity": "2.13571758343362928056",
        "sequenceIndexInSide": 8,
        "side": "BUY",
        "status": "done",
        "type": "LIMIT",
      },
      {
        "id": 10,
        "oco": {
          "isOco": false,
        },
        "price": "50",
        "quantity": "2.46675380886584181904",
        "sequenceIndexInSide": 9,
        "side": "BUY",
        "status": "done",
        "type": "LIMIT",
      },
      {
        "id": 11,
        "oco": {
          "isOco": false,
        },
        "price": "95.95",
        "quantity": "0.83689026279428100526",
        "sequenceIndexInSide": 0,
        "side": "SELL",
        "status": "canceled",
        "type": "LIMIT",
      },
      {
        "id": 12,
        "oco": {
          "isOco": false,
        },
        "price": "93.29525691699604743083",
        "quantity": "1.76444363739127578609",
        "sequenceIndexInSide": 1,
        "side": "SELL",
        "status": "canceled",
        "type": "LIMIT",
      },
      {
        "id": 13,
        "oco": {
          "isOco": false,
        },
        "price": "90.54896693911690903488",
        "quantity": "2.79566474209028763066",
        "sequenceIndexInSide": 2,
        "side": "SELL",
        "status": "canceled",
        "type": "LIMIT",
      },
      {
        "id": 14,
        "oco": {
          "isOco": false,
        },
        "price": "87.70674325863052500428",
        "quantity": "3.94612078702012271976",
        "sequenceIndexInSide": 3,
        "side": "SELL",
        "status": "canceled",
        "type": "LIMIT",
      },
      {
        "id": 15,
        "oco": {
          "isOco": false,
        },
        "price": "84.76357671520760958221",
        "quantity": "5.23463155734153801955",
        "sequenceIndexInSide": 4,
        "side": "SELL",
        "status": "canceled",
        "type": "LIMIT",
      },
      {
        "id": 16,
        "oco": {
          "isOco": false,
        },
        "price": "81.71367022001670668914",
        "quantity": "6.68420617395313023182",
        "sequenceIndexInSide": 5,
        "side": "SELL",
        "status": "canceled",
        "type": "LIMIT",
      },
      {
        "id": 17,
        "oco": {
          "isOco": false,
        },
        "price": "78.55021512879395239821",
        "quantity": "8.32334054812162296415",
        "sequenceIndexInSide": 6,
        "side": "SELL",
        "status": "canceled",
        "type": "LIMIT",
      },
      {
        "id": 18,
        "oco": {
          "isOco": false,
        },
        "price": "75.26508310620099002459",
        "quantity": "10.18785589873828344718",
        "sequenceIndexInSide": 7,
        "side": "SELL",
        "status": "canceled",
        "type": "LIMIT",
      },
      {
        "id": 19,
        "oco": {
          "isOco": false,
        },
        "price": "71.84839153458161975409",
        "quantity": "12.32357348217191272774",
        "sequenceIndexInSide": 8,
        "side": "SELL",
        "status": "canceled",
        "type": "LIMIT",
      },
      {
        "id": 20,
        "oco": {
          "isOco": true,
          "ocoType": "limit",
        },
        "price": "68.28787356260957689196",
        "quantity": "14.79032729103775454678",
        "sequenceIndexInSide": 9,
        "side": "SELL",
        "status": "active",
        "type": "LIMIT",
      },
      {
        "id": 21,
        "oco": {
          "isOco": true,
          "ocoType": "stop-limit",
          "stopPrice": "45",
        },
        "price": "45",
        "quantity": "14.79032729103775454678",
        "sequenceIndexInSide": 9,
        "side": "SELL",
        "status": "active",
        "type": "STOP_LIMIT",
      },
    ],
  },
  "isCycleOver": false,
}
`;
